specs/spec.md

```diff
@@ -0,0 +1,329 @@
+# @TODO
+- [ ] Link to parsers.
+- [ ] attestations
+- [ ] what else needs to be defined in the transport header?
+
+# Hobbits
+
+*Written by Dean Eigenmann <dean@status.im> & Rene Nayman <rene@whiteblock.io>* 
+
+## Table of Contents
+
+1. [Abstract](#abstract)
+2. [Flow](#flow)
+3. [Wire Protocol](#wire-protocol)
+    - [Messages](#messages)
+4. [Transport](#transport)
+5. [Protocols](#protocols)
+    - [RPC](#rpc)
+    - [Gossip](#gossip)
+    - [PING](#ping)
+6. [Implementations](#implementations)
+
+## Abstract
+
+Hobbits is a modular wire protocol which allows implementers to experiment with various application logic over a practical network in an agnostic and expedient manner.
+
+Hobbits wire protocol was developed so that Eth2.0 clients could begin network-level testing without requiring libp2p.
+
+## Flow
+
+We layout a network where the nodes are peered in the following manner:
+
+![nodes](./nodes.png)
+
+The following flow dictates how blocks are shared amongst peers:
+
+![sequence diagram](./diagram.png)
+
+1. Node 1 receives a new block `0xCAB` and gossips `hash(0xCAB)` to Node 2 and Node 3
+2. Node 2 requests the block bodies for block `0xCAB` from Node 1 in an RPC Command query
+3. Node 1 responds to Node 2 with the block bodies for block `0xCAB` in an RPC Command response
+4. Node 2 receives a new block `0xCAB` and gossips it to Node 1 and Node 3
+
+## Wire Protocol
+
+### Messages
+
+The `message` format looks as follows:
+
+```
+EWP <version> <protocol> <header-length> <body-length>
+<header><body>
+```
+
+**NOTE:** Message`body`s along with `header` fields are BSON encoded.
+
+#### Fields
+
+Every hobbit message contains the following fields: 
+
+| Field | Definition | Validity |
+|:------:|----------|:----:|
+| `version` | Defines the EWP version number e.g. `0.2`. | `(\d+\.)(\d+)` |
+| `protocol` | Defines the [protocol](#protocols). | `(RPC\|GOSSIP\|PING)` |
+| `header` | Defines the header | payload |
+| `body` | Defines the body | payload |
+
+## Transport
+
+A Hobbits node listens for and sends out messages. 
+
+Hobbits packets are sent as follows: 
+
+```
+<length><packet>
+```
+
+- **Length**: `big endian` encoded `uint32` that represents the length of the packet, with a length of 4 bytes.
+- **Packet**: Byte encoded string of the hobbits message.
+
+## Protocols
+
+Hobbits defines 3 protocols that dictate how messages are interpreted and responded to.
+
+### RPC
+
+<!-- @TODO WRITE SOME SHIT -->
+
+#### Envelope
+
+```python
+{ 
+  'method_id': 'uint16' ## byte representing the method
+  'id': 'uint64' ## id of the request
+  'body': 'bytes' ## body of the request itself
+}
+```
+
+The `body` field contains an [SSZ](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/simple-serialize.md) encoded byte array. 
+
+##### Example
+
+Below you will find an example `RPC` request.
```

Comment:
isn't this a respond? I thought `bodies` contains an ssz encoded byte away

Reply:
OK should be message instead.

Reply:
```suggestion
Below you will find an example `RPC` message.
```